File: 01-arrays_&_hashing/00001-Two_Sum/00001-two_sum.java
// 1: Two Sum
// https://leetcode.com/problems/two-sum/

import java.util.HashMap;
import java.util.Map;

class Solution {
    // SOLUTION
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<Integer, Integer>();

        for (int i=0; i<nums.length; map.put(nums[i], i++))
            if (map.containsKey(target - nums[i]))
                return new int[] {map.get(target - nums[i]), i};
        
        return new int[] {0,0};
    }

    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        int[] nums = {3,2,4};
        int target = 6;

        // OUTPUT
        var result = o.twoSum(nums, target);
        System.out.println("[" + result[0] + ", " + result[1] + "]");
    }
}

File: 01-arrays_&_hashing/00001-Two_Sum/00001-two_sum.go
// 1: Two Sum
// https://leetcode.com/problems/two-sum/

package main

import "fmt"

// SOLUTION
func twoSum(nums []int, target int) []int {
	tmap := make(map[int]int)

	for i, n := range nums {
		_, m := tmap[n]

		if m {
			return []int{tmap[n], i}
		} else {
			tmap[target-n] = i
		}
	}

	return nil
}

func main() {
	// INPUT
	nums := []int{3,2,4}
	target := 6

	// OUTPUT
	result := twoSum(nums, target)
	fmt.Printf("[%d, %d]\n", result[0], result[1])
}
